language/move-vm/integration-tests/src/tests/exec_func_effects_tests.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use std::convert::TryInto;
+
+use move_core_types::{
+    account_address::AccountAddress,
+    identifier::Identifier,
+    language_storage::ModuleId,
+    value::{MoveValue, serialize_values}, vm_status::StatusCode
+};
+use move_vm_runtime::{move_vm::MoveVM, session::ExecutionResult};
+use move_vm_test_utils::InMemoryStorage;
+use move_vm_types::gas_schedule::GasStatus;
+
+use crate::compiler::{compile_units, as_module};
+
+
+const TEST_ADDR: AccountAddress = AccountAddress::new([42; AccountAddress::LENGTH]);
+const TEST_MODULE_ID: &str = "M";
+const EXPECT_MUTREF_OUT_VALUE: u64 = 90;
+const USE_MUTREF_LABEL: &str = "use_mutref";
+const USE_REF_LABEL: &str = "use_ref";
+const FUN_NAMES: [&str; 2] = [USE_MUTREF_LABEL, USE_REF_LABEL];
+
+// ensure proper errors are returned when ref & mut ref args fail to deserialize
+#[test]
+fn fail_arg_deserialize() {
+    let mod_code = setup_module();
+    // all of these should fail to deserialize because the functions expect u64 args
+    vec![MoveValue::U8(16), MoveValue::U128(512), MoveValue::Bool(true)]
+    .iter()
+    .for_each(|mv| {
+        FUN_NAMES.iter().for_each(|name| {
+            match run(&mod_code, name, mv.clone()) {
+                ExecutionResult::Success { .. } => {
+                    panic!("Should have failed to deserialize non-u64 type to u64");
+                },
+                ExecutionResult::Fail { error, .. } => {
+                    assert_eq!(error.major_status(), StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT);
+                }
+            }
+        });
+    });
+}
+
+// check happy path for writing to mut ref args - may be unecessary / covered by other tests
```

Comment:
What i really want to test is the new error handling for mutable ref serialization, but it seems tricky to get the Move code to return a value that fails to serialize. If I can figure out how to cause that error, i'll add the test for it to mirror this.

Reply:
The Move code shouldn't return a value that fails to serialize unless there's a bug in the VM, so this would indeed be tricky without some kind of fault injection... This test looks great, thanks!